agent/config/builder.go

```diff
@@ -2473,3 +2452,75 @@ func validateAbsoluteURLPath(p string) error {
 
 	return nil
 }
+
+func (b *builder) buildTLSConfig(rt RuntimeConfig, t TLS) (tlsutil.Config, error) {
+	var c tlsutil.Config
+
+	// Consul makes no outgoing connections to the public gRPC port (internal gRPC
+	// traffic goes through the multiplexed internal RPC port) so return an error
+	// rather than let the user think this setting is going to do anything useful.
+	if t.GRPC.VerifyOutgoing != nil {
+		return c, errors.New("verify_outgoing is not valid in the tls.grpc stanza")
+	}
+
+	// Similarly, only the internal RPC configuration honors VerifyServerHostname
+	// so we call it out here too.
+	if t.Defaults.VerifyServerHostname != nil || t.GRPC.VerifyServerHostname != nil || t.HTTPS.VerifyServerHostname != nil {
+		return c, errors.New("verify_server_hostname is only valid in the tls.internal_rpc stanza")
+	}
+
+	// TLS is only enabled on the gRPC listener if there's an HTTPS port configured
+	// for historic and backwards-compatibility reasons.
+	if rt.HTTPSPort <= 0 && (t.GRPC != TLSListener{}) {
+		b.warn("tls.grpc was provided but TLS will NOT be enabled on the gRPC listener without an HTTPS listener configured (e.g. via ports.https)")
+	}
```

Comment:
For posterity: we considered decoupling the gRPC TLS "enabledness" from whether there's an HTTPS port configured, but it turns out there's a lot of subtle complexity hidden here (particularly when considering backwards compatibility and auto-config).

I thought we should at least warn the user if their `tls.grpc` settings will be ineffectual, because this tripped me up while testing!